File: codes/backtraking/40.combination-sum-ii/combination-sum-ii.cpp
#include <vector>
#include <algorithm>
using namespace std;
class Solution {
public:
    vector<vector<int>> result;
    vector<int> path;
    using vit = vector<int>::const_iterator;
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        backtracing(candidates.cbegin(), candidates.cend(), target);
        return result;
    }
    void backtracing(vit beg, vit end, int target) {
        if (beg == end) {
            return;
        }
        for (auto it = beg; it != end; it++) {
            // 同层去重，而树枝上不去重
            // 同层相当于第 i 个位置上，第 i 个位置上的元素是不能重复的，这依赖于数组是排序的。
            if (it > beg && *it == *(it-1)) {
                continue;
            }
            auto val = target - *it;
            if (val < 0) {
                return;
            }
            path.push_back(*it);
            backtracing(it + 1, end, val);
            if (val == 0) {
                result.push_back(path);
            }
            path.pop_back();
        }
    }
};

File: codes/backtraking/40.combination-sum-ii/combination-sum-ii.go
package main

import "sort"

func combinationSum2(candidates []int, target int) [][]int {
	var result [][]int
	var path []int
	var backtracing func(int, int)
	backtracing = func(beg, target int) {
		if beg >= len(candidates) {
			return
		}
		for i := beg; i < len(candidates); i++ {
			if i != beg && candidates[i] == candidates[i-1] {
				continue
			}
			target -= candidates[i]
			if target < 0 {
				return
			}
			path = append(path, candidates[i])
			if target == 0 {
				result = append(result, append([]int{}, path...))
				path = path[:len(path)-1]
				return
			}

			backtracing(i+1, target)
			path = path[:len(path)-1]
			target += candidates[i]
		}
	}

	sort.Ints(candidates)
	backtracing(0,target)
	return result

}
